helix-core/src/main/java/org/apache/helix/controller/stages/PerReplicaThrottleStage.java

```diff
@@ -220,6 +271,274 @@ private MessageOutput throttlePerReplicaMessages(IdealState idealState,
+   *  of the toState, we will treat it as Load, otherwise, it is Recovery.
+   *
+   *  Note, there can be customerized model that having more than one route to a top state. For
+   *  example S1, S2, S3 are three levels of states with S1 as top state with lowest priority.
+   *  It is possible that S2 can transits upward to S1 while S3 can also transits upward to S1.
+   *  Thus, we consider S1 meet count requirement of both S2 and S3. In propagation time, we will
+   *  add state count of S1 to both S2 and S3.
+   */
+  private void propagateCountsTopDown(StateModelDefinition stateModelDef,
+      Map<String, Integer> stateCountMap) {
+    List<String> stateList = stateModelDef.getStatesPriorityList();
+    if (stateList == null || stateList.size() <= 0) {
+      return;
+    }
+
+    Map<String, Integer> statePriorityMap = new HashMap<>();
+
+    // calculate rank of each state. Next use the rank to compare if a transition is upward or not
+    int rank = 0;
+    for (String state : stateList) {
+      statePriorityMap.put(state, Integer.valueOf(rank));
+      rank++;
+    }
+
+    // given a key state, find the set of states that can transit to this key state upwards.
+    Map<String, Set<String>> fromStatesMap = new HashMap<>();
+    for (String transition : stateModelDef.getStateTransitionPriorityList()) {
+      // Note, we assume stateModelDef is properly constructed.
+      String[] fromStateAndToState = transition.split("-");
+      String fromState = fromStateAndToState[0];
+      String toState = fromStateAndToState[1];
+      Integer fromStatePriority = statePriorityMap.get(fromState);
+      Integer toStatePriority = statePriorityMap.get(toState);
+      if (fromStatePriority.compareTo(toStatePriority) <= 0) {
+        // skip downward transitition
+        continue;
+      }
+      fromStatesMap.putIfAbsent(toState, new HashSet<>());
+      fromStatesMap.get(toState).add(fromState);
+    }
+
+    // propagation by adding state counts of current state to all lower priority state that can
+    // transit to this current state
+    int index = 0;
+    while (true) {
+      if (index == stateList.size() - 1) {
+        break;
+      }
+      String curState = stateList.get(index);
+      String num = stateModelDef.getNumInstancesPerState(curState);
+      if ("-1".equals(num)) {
+        break;
+      }
+      stateCountMap.putIfAbsent(curState, 0);
+      Integer curCount = stateCountMap.get(curState);
+      // for all states S that can transition to curState, add curState count back to S in stateCountMap
+      for (String fromState : fromStatesMap.getOrDefault(curState, Collections.emptySet())) {
+        Integer fromStateCount = stateCountMap.getOrDefault(fromState, 0);
+        stateCountMap.put(fromState, Integer.sum(fromStateCount, curCount));
+      }
+      index++;
+    }
+  }
+
+  Map<String, Integer> getPartitionExpectedStateCounts(Partition partition,
```

Comment:
It seems that the only difference between getPartitionExpectedStateCounts and getPartitionCurrentStateCounts is that you need to calculate the expected state count map from the preference list. Is it possible to move this logic outside to calculateExpectedStateCounts so that we can avoid some duplicate code?

Reply:
The original code has calculatedExpectedStateCounts and calculateCurrentStateCount in one function. Later, based on the review feedback from you and @pkuwm, we decided to separate this into two paths. That is why I would say let stick to current way and not moving back.

The second point is that the code getPartitionExpectedStateCounts and getPartitionCurrentStateCounts does not have much duplication, seem to me? Do you have a proposal of how to de-dupe the duplicate logic?

To be on the same page, let me list the code here. Maybe you can draw some code to illustrate the idea.

```
  private Map<String, Integer> getPartitionExpectedStateCounts(Partition partition,
      Map<String, List<String>> preferenceLists, StateModelDefinition stateModelDef,
      IdealState idealState, Set<String> enabledLiveInstance) {
    Map<String, Integer> expectedStateCountsOut = new HashMap<>();

    List<String> preferenceList = preferenceLists.get(partition.getPartitionName());
    if (preferenceList == null) {
      preferenceList = Collections.emptyList();
    }

    int replica =
        idealState.getMinActiveReplicas() <= 0 ? idealState.getReplicaCount(preferenceList.size())
            : idealState.getMinActiveReplicas();
    Set<String> activeList = new HashSet<>(preferenceList);
    activeList.retainAll(enabledLiveInstance);

    // For each state, check that this partition currently has the required number of that state as
    // required by StateModelDefinition.
    LinkedHashMap<String, Integer> expectedStateCountMap =
        stateModelDef.getStateCountMap(activeList.size(), replica);

    expectedStateCountsOut.putAll(expectedStateCountMap);
    propagateCountsTopDown(stateModelDef, expectedStateCountsOut);

    return expectedStateCountsOut;
  }
```

and 

```
private Map<String, Integer> getPartitionCurrentStateCounts(Map<String, String> currentStateMap,
      String resourceName, String partitionName, StateModelDefinition stateModelDef,
      ResourceControllerDataProvider cache) {
    // Current counts without disabled partitions or disabled instances
    Map<String, String> currentStateMapWithoutDisabled = new HashMap<>(currentStateMap);
    currentStateMapWithoutDisabled.keySet()
        .removeAll(cache.getDisabledInstancesForPartition(resourceName, partitionName));
    Map<String, Integer> currentStateCountsOut =
        StateModelDefinition.getStateCounts(currentStateMapWithoutDisabled);

    propagateCountsTopDown(stateModelDef, currentStateCountsOut);

    return currentStateCountsOut;
  }

```

Reply:
Let me take a look and reply with a proposal. Please don't resolve the conversion with an open discussion.

Reply:
If you were referring to the very first version, then the original method is called "getPartitionExpectedAndCurrentStateCountMap". That method is definitely NOT what I'm trying to suggest you to do. That was one single method, but you were just putting all the logic together.

For this new code, what I thought is actually avoid duplicating the loops in the calculateXXXXCount() methods. However, as you also mentioned, you will have to pre-calculate the input with a loop for expected state map and current state map separately anyway, so it is not possible to save here.

The only possible way is to abstract the pre-calculate logic into other classes and then make it uniform calls. This requires a larger scope refactoring and may still not make the cut. So obviously, the cost is much higher than the gain. Let's just keep the current implemenation. I can help to have a try to see if it can be refined, later once this PR is in.